### PR TITLE
Remove `AS::Multibyte`'s unicode table

### PR DESCRIPTION
**Pull request by mtsmfm**
### Summary

@amatsuda tells us `AS::Multibyte` has big Unicode table and ruby may have similar feature (https://speakerdeck.com/a_matsuda/3x-rails?slide=156).

So I investigated and noticed that we can remove the table if we accept following changes.
- Supported Unicode version will depend on Ruby
  - Ruby 2.2 supports Unicode 7.0.0
  - Ruby 2.3 supports Unicode 8.0.0
  - Ruby 2.4 supports Unicode 9.0.0
- Unicode case mappings doesn't work until Ruby 2.4
- ~~Grapheme doesn't work perfect~~ Ruby 2.4 works perfect! https://bugs.ruby-lang.org/issues/12831

If we merge this PR, we'll have following benefits:
- Reduce memory usage(about 1M)
- Reduce repo/gem size(about 1M)
- Reduce code lines(about 400 lines)

I'll show you the way I tried to remove the table and why the changes will cause.

---

`AS::Multibyte::Unicode`'s features are divided into 4 main groups.
1. normalize
   -   [normalize](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L285)
   -   [compose](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L180)
   -   [decompose](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L159)
   -   [reorder_characters](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L143) (this is public method but probably for compose/decompose only)
2. case mapping
   -   [upcase](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L307)
   -   [downcase](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L303)
   -   [swapcase](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L311)
3. pack/unpack grapheme
   - [pack_graphemes](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L138)
   - [unpack_graphemes](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L80)
4. tidy bytes
   - [tidy_bytes](https://github.com/rails/rails/blob/v5.0.0.1/activesupport/lib/active_support/multibyte/unicode.rb#L245)

1 ~ 3 use unicode database.
So let remove them.
#### normalize

Since 2.2, Ruby has `String#unicode_normalize`.

https://www.ruby-lang.org/en/news/2014/12/25/ruby-2-2-0-released/
https://bugs.ruby-lang.org/issues/10084

[It has similar option](https://ruby-doc.org/stdlib-2.2.5/libdoc/unicode_normalize/rdoc/String.html#method-i-unicode_normalize) so it seems that we can simply replace it.
(I tried https://github.com/rails/rails/pull/26403)
But Ruby 2.2 supports Unicode 7.0.0 and Ruby 2.3 supports 8.0.0.
`AS::Multibyte::Unicode` has its own `UNICODE_VERSION` but now it depends on Ruby's version.
#### case mappings

[Ruby 2.4 supports Unicode case mappings](https://www.ruby-lang.org/en/news/2016/09/08/ruby-2-4-0-preview2-released/).
But Ruby 2.2 and 2.3 don't.
#### pack/unpack grapheme

Since 2.0, [Ruby changed its regexp engine](https://www.ruby-lang.org/en/news/2013/02/24/ruby-2-0-0-p0-is-released/).
So we can split grapheme with `/X/`.
(ex. `"\u304B\u3099\u304C\u3099\u304E\u3099".scan(/\X/)`)

~~But [it's grapheme cluster is not "true"](https://github.com/k-takata/Onigmo/issues/46).~~
~~So some tests will fail.~~
@nurse fixed on Ruby 2.4!
https://bugs.ruby-lang.org/issues/12831

#### Memory

Unicode database consume 1.5 M memory.

``` rb
require 'active_support/core_ext/string/multibyte'
require 'objspace'

# load unicode database
''.mb_chars.upcase

db = ActiveSupport::Multibyte::Unicode.send(:database)
db_memsize = db.instance_variables.map do |ivar|
  ObjectSpace.memsize_of(db.instance_variable_get(ivar))
end.inject(:+)

puts "db_memsize: #{db_memsize} Bytes"
#=> db_memsize: 1501184 Bytes
```

Ruby's Unicode database is smaller than `AS::Multibyte::Unicode::UnicodeDatabase`

``` rb
require 'objspace'

# load unicode database
''.unicode_normalize

db_memsize = UnicodeNormalize.constants.map do |const|
  ObjectSpace.memsize_of(UnicodeNormalize.const_get(const))
end.inject(:+)

puts "db_memsize: #{db_memsize} Bytes"
#=> db_memsize: 404558 Bytes
```

### CODE REVIEW

**Comment by rafaelfranca** [comment]
While I like the idea but I don't think we should remove some features for
the supported Ruby versions because of 1mb of memory. I can see we merging
it when we drop support to Ruby < 2.4 but while we need to support 2.2 and
2.3 I think this extra memory usage is fine

**Comment by amatsuda** [comment]
@mtsmfm Thank you for the PR and summary! This is very well done!
I suppose another benefit that we can get from this change is maybe 100-200ms boot time speed up for not loading and parsing 1MB .dat file.

@rafaelfranca Agreed. Actually we (@mtsmfm and I) already talked about that in person, so he understands that this PR might not be fully merged right now (and maybe we need to wait until Rails 6).
But still, he wanted to show us the goal of this small project, and I think he did it perfectly :)

So the next step should be to find out what we can cherry-pick to 5.1 branch from this huge PR.

**Comment by mtsmfm** [comment]
@amatsuda Thank you for your following up!

> So the next step should be to find out what we can cherry-pick to 5.1 branch from this huge PR.

I think it is difficult to cherry-pick 
because `AS::Multibyte::Unicode` has its own `UNICODE_VERSION`.
So we must replace all methods with Ruby's feature or do nothing :cry:

Should I close this PR at this time or keep opening until rails 6?

**Comment by jeremy** [comment]
Since the UnicodeDatabase lazy loads, perhaps we could leave it as-is for older Rubies but start switching to native implementation for Ruby 2.4+?

**Comment by mtsmfm** [comment]
@jeremy Sorry, I don't understand what "switching to native implementation for Ruby 2.4+" means :bowing_man: 
Do you mean "we should implement new API such like `AS::Multibyte2` and deprecate `AS::Multibyte` :question: 

> Since the UnicodeDatabase lazy loads

Ruby 2.4+ also lazy loads unicode normalize database.
https://github.com/ruby/ruby/blob/v2_4_0/lib/unicode_normalize.rb#L33
(And I can't find where ruby loads case mapping database :sweat_drops: )

**Comment by mtsmfm** [comment]
I think it's time to merge because Rails 6 requires Ruby 2.4+

**Comment by mtsmfm** [comment]
@jeremy @amatsuda Could you review?

**Comment by rafaelfranca** [comment]
Can you also remove `bin/generate_tables`?

**Comment by rafaelfranca** [comment]
And the `unicode_tables.dat` file

**Comment by mtsmfm** [comment]
Oops, removed and force pushed!

**Comment by jeremy** [comment]
🎉🎉🎉

**Comment by amatsuda** [comment]
👍 👍 👍

**Comment by mtsmfm** [comment]
:joy:

**Comment by matthewd** [comment]
🎉

Should we consider deprecating some of these APIs now? I see a lot are trivial wrappers for core methods.

**Comment by rafaelfranca** [comment]
👍 to deprecate some of those wrappers

File: activesupport/lib/active_support/multibyte/unicode.rb [hunk 1]
```diff
@@ -266,129 +101,37 @@ def tidy_bytes(string, force = false)
...
 
       def downcase(string)
-        apply_mapping string, :lowercase_mapping
+        string.downcase
       end
 
       def upcase(string)
-        apply_mapping string, :uppercase_mapping
+        string.upcase
       end
```
[the remark this answers is not in the document]
Comment (mtsmfm) [reply activesupport/lib/active_support/multibyte/unicode.rb]: I've just noticed `"\uFB03"` will be different...

[the remark this answers is not in the document]
Comment (mtsmfm) [reply activesupport/lib/active_support/multibyte/unicode.rb]: Let me think a while

**Comment by mtsmfm** [comment]
@jeremy @rafaelfranca @amatsuda 

Sorry, I've found this change breaks previous behavior in some edge cases 🙇 

```
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  gem "activesupport", "5.2.1"
end

require "active_support/all"

str = "\uFB03"
as_result = ActiveSupport::Multibyte::Unicode.upcase(str)
mri_result = str.upcase

puts as_result # => ﬃ
puts mri_result # => FFI

p as_result.codepoints # => [64259]
p mri_result.codepoints # => [70, 70, 73]
```

It seems the previous implementation doesn't follow [special casting](https://www.unicode.org/Public/11.0.0/ucd/SpecialCasing.txt) but Ruby does.

I think it's a bug and Ruby is correct so it's ok to change this behavior though.

What do you think?

**Comment by rafaelfranca** [comment]
It seems fine to me too.

**Comment by mtsmfm** [comment]
So then, I think it's also ok to change `Chars#capitalize` @frodsan 

https://github.com/rails/rails/pull/26743#discussion_r223622214

**Pull request merged**